engine/function/src/templates/Cast.ftl

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright (c) 2016-2021 Deephaven Data Labs and Patent Pending
+ */
+
+package io.deephaven.function;
+
+import io.deephaven.vector.*;
+
+import static io.deephaven.util.QueryConstants.*;
+
+/**
+ * Functions for casting between types.
+ */
+@SuppressWarnings("unused")
+public class Cast {
+
+    /**
+     * Cast can not preserve the value.
+     */
+    public static class CastDoesNotPreserveValue extends UnsupportedOperationException {
+
+        private static final long serialVersionUID = 5998276378465685974L;
+
+        /**
+         * Creates a new exception.
+         *
+         * @param message error message.
+         */
+        public CastDoesNotPreserveValue(final String message) {
+            super(message);
+        }
+    }
+
+    private static boolean isLosingPrecisionDouble(final long v) {
```

Comment:
I think we need a more precise test here. There are lots of long values greater than 9007199254740992 that convert to double without losing precision. For example all the powers of two. One example is 2^63.

Maybe a test like this would work: `v != (long)(double)v`
but I'm not sure about the edge cases

Reply:
hmmm.... When I implemented this originally, I was just looking at a range of acceptable values.  I hadn't considered the powers of two, or anything like that.  I can convince myself that either the range case or the `v != (long)(double)v` is the "right one".  I made the suggested change, but I'm not 100% convinced it is the right way to go... and I'm not 100% convinced it is the wrong way to go.  Thoughts?

Reply:
I'm kind of torn myself really. I'm really not sure any more.

I think maybe the *original* way you had it might be less surprising to the user, because it represents a dense range of acceptable values from -9007199254740992L to 9007199254740992L, where every value in the range is legal.

So maybe go back to the way you had it, and if further enlightenment strikes, we can revisit the issue later?